Add forgejo vcs provider

Infinoid opened the pull request:
This adds forgejo as a VCS provider.  I've been developing and testing it against a local forgejo 11.0.1 instance.

It probably works with Gitea, too, since the APIs are largely compatible.

TODO:
* [x] get it to the point where it can apply a basic `random_password` resource
* [ ] see it working in a larger workspace, with variables, state storage, modules, etc
* [ ] per-instance configuration of API URLs (forgejo is self-hosted and federated, use of multiple instances is likely)
* [ ] run the actual test suite on it (my `podman` environment is very confused by the `compose` stuff in the testsuite)
* [ ] Parse and return the `user.html_url` field for Author URLs (needs a [small fix](https://codeberg.org/mvdkleijn/forgejo-sdk/pulls/57) to the forgejo SDK)
* [ ] documentation
    * [ ] min/max versions
    * [ ] vcs setup
    * [ ] provide an example of the oidc setup, too, just to reduce guesswork
* [ ] test gitea
    * [ ] min/max versions

Infinoid commented:
This is something I threw together in the last day or so.  It's very unpolished, but it's starting to seem like something that should actually work.

Reviews/guidance/suggestions are welcome.

Infinoid commented:
> Keep going, and feel free to raise PRs that refactor the codebase to make your task easier.

Great!  Will do.  The fact that it can already list repos, attach a workspace to a repo from the list, receive webhooks, make a plan, and apply it is quite encouraging.

This is a weekend project for me, so that's when most activity should happen.

The `Client` interface [definition](https://github.com/leg100/otf/blame/94a474e8040dd65be8139570f50d52a6c6fffb15/internal/vcs/client.go#L8) has some nice comments which let me get pretty far with it.  But there are still some things I'm trying to figure out.  Like:
* should `ListRepositories` return the *writable* repos?  (That's my assumption so far; I imagine the list of *readable* git repos on github would be quite long...)
* if every `GetRepoTarball` implementation needs to remove a top-level folder from the archived files, wouldn't it be easier to unpack it with `tar --strip-components=1`?
* Is it okay if `GetWebhook` returns a different set of events from what was passed to `CreateWebhook`?  "Tag" events aren't really a thing; pushing a tag results in a `push` event, so I've kinda merged them together.
* When parsing webhooks, how much do I need to care about the myriad of `pull_request` [event subtypes](https://forgejo.org/docs/latest/user/actions/#onpull_request)?  Does it just need code changes, like `opened`, `reopened` and `synchronize` events, or should it match labels and comments and such too?

Thanks.

leg100 commented:
> The Client interface [definition](https://github.com/leg100/otf/blame/94a474e8040dd65be8139570f50d52a6c6fffb15/internal/vcs/client.go#L8) has some nice comments which let me get pretty far with it. But there are still some things I'm trying to figure out. Like:

If in doubt, treat the Github implementation as the reference.
> 
> should ListRepositories return the writable repos? (That's my assumption so far; I imagine the list of readable git repos on github would be quite long...)

According to the [docs for the upstream Github endpoint](https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-repositories-for-the-authenticated-user), it lists repos the user has *explicit* permission to access.

> if every GetRepoTarball implementation needs to remove a top-level folder from the archived files, wouldn't it be easier to unpack it with tar --strip-components=1?

What, literally call out to `tar --strip-components=1`? I think that's a bit naff (requires external executable `tar`, slower, etc). I agree though that if every impl is doing the same thing then perhaps have `Unpack()` perform the equivalent in Go.

> Is it okay if GetWebhook returns a different set of events from what was passed to CreateWebhook? "Tag" events aren't really a thing; pushing a tag results in a push event, so I've kinda merged them together.

I think that's okay, yes.

> When parsing webhooks, how much do I need to care about the myriad of pull_request [event subtypes](https://forgejo.org/docs/latest/user/actions/#onpull_request)? Does it just need code changes, like opened, reopened and synchronize events, or should it match labels and comments and such too?

No, I don't think it needs to match labels and comments etc. It's about what behaviour the user would reasonably expect from the integration, and I don't think adding something like a comment should trigger runs (although that may change should we implement Atlantis type comment actions).

Infinoid commented:
> What, literally call out to `tar --strip-components=1`? I think that's a bit naff (requires external executable `tar`, slower, etc). I agree though that if every impl is doing the same thing then perhaps have `Unpack()` perform the equivalent in Go.

Oh, I was referring to semantics, not implementation.  You're right, calling `tar` directly is a bad idea.  (Also,  there are other `tar`s than the GNU one, and they vary wildly.)

I had just noticed that every VCS provider does the same thing, re-packing the archive in a temp folder to strip exactly 1 leading folder from the files.  Future VCS providers might differ, but each VCS provider will know what to expect, so it could return the number of leading folders along with the archive itself.

Moving it out of the VCS provider means it could be handled with `os.Chdir()` instead of all the extra I/O ops.  It's also better DRY.

That was the thought process, but it's beyond the scope of this PR.

Infinoid commented:
Ok!  This is all working pretty well.  I set up a [test repo](https://v11.next.forgejo.org/infinoid/test) on forgejo's test/demo instance, I stood up a temporary OTF installation, set up the VCS provider, connected the workspace to the repo and it all works.  I was able to:

* create runs from the dropdown menu on the workspace page (plan-only and plan-and-apply)
* push a commit to the repo's main branch and see it create a run
* create a PR and see it create a (plan-only) run
* merge the PR and see it create a run

Here's the resulting run list:

![image](https://github.com/user-attachments/assets/c574b43a-e4ee-4f07-b3e8-655f2b2eca25)

There are a couple of hacks I need to clean up, though.  And a few questions:

## client object in webhook handler?

I'd like to make calls to a `forgejo.Client` object from within the webhook handler.  Is there a good way to get that?

The reason is, the list of changed files isn't part of forgejo's webhook [json data](https://github.com/Infinoid/otf/blob/8fb3cfc046042ea60189355bd7b529a1f283428b/internal/forgejo/testdata/pr_opened.json).  But if the webhook handler leaves it empty, then later on, the UI template panics when it [tries to slice](https://github.com/Infinoid/otf/blob/8fb3cfc046042ea60189355bd7b529a1f283428b/internal/run/view.templ#L119) the first 7 characters of a nonexistent commit SHA.

I solve this by making [an API call](https://github.com/Infinoid/otf/blob/8fb3cfc046042ea60189355bd7b529a1f283428b/internal/forgejo/pr_event.go#L67) to populate the list of changed files.  I've hacked the forgejo provider to [stash](https://github.com/Infinoid/otf/blob/8fb3cfc046042ea60189355bd7b529a1f283428b/internal/forgejo/client.go#L33) client objects that it creates, so that the webhook handler can use it, but this is obviously a hack.  And not a reliable one - if the webhook arrives before the first VCS client object is called, the handler can't look up the files and is forced to return incomplete data.

What's the right way to make this happen?  Passing in a `forgejo.Client` object would not work through the generic interface, but a `vcs.Client` would.  Or alternately, passing in a `vcs.NewTokenClientOptions` object would be enough, so it could create its own if necessary.  Either of those would work for me, but I'm open to other ideas.

## logging?

It's been useful to log some debug messages as I develop this.  However, the vcs provider plugin interface is standalone, and rather isolated from the rest of otfd.  It doesn't have the logging infrastructure that the rest of otfd does.

Does it make sense to add a `logr.Logger` field to the `vcs.NewTokenClientOptions` struct?  I think that without it, if I made my own Logger object, it wouldn't be configured properly with otfd's verbosity setting, right?

Is there a better way to do this?  Currently I'm using Go's builtin `log.Printf` for debugging, and it's enough for me to develop with, but it's not very nice.

## SetStatus is never called?

As shown above, it all seems to be working pretty well for me.  However, I noticed that the `client.SetStatus` method is never called.  Should it be?  I see [the code to call it](https://github.com/Infinoid/otf/blob/8fb3cfc046042ea60189355bd7b529a1f283428b/internal/run/reporter.go#L144-L151), but it's not actually happening for me.  Does OTF normally call SetStatus, and I'm just misconfiguring it somehow?

Infinoid commented:
I also wrote a [quick guide](https://github.com/Infinoid/otf/blob/feature/forgejo-vcs/docs/docs/forgejo_setup.md) on how to set things up.  I don't know how that should be organized, but I hope this is a starting point.

leg100 commented:
Logging: You'll find logging two places: in service endpoints and in "subsystems". Service endpoints can be called by different presentation layers (API, UI, CLI, etc), so logging in the endpoint makes sense rather than in each and every presentation layer. And the subsystems are autonomous daemons that react mostly to events and it's helpful for users to know what they're up to.

Whereas I think placing logging in various VCS clients (GitHub, Forgejo, etc) will result in a lot of duplicate logging. The logging is better placed in the thing that calls the client interface, e.g. the VCS reporter subsystem.

Indeed it is the reporter that calls `SetStatus()`. Perhaps you can add some logging there? As for why it's not being called, I don't know, but it's worth checking the logic of the reporter's `handleRun()` method and its unit tests. Again, adding further logging in the reporter may help work out what's going on.

leg100 commented:
> I also wrote a [quick guide](https://github.com/Infinoid/otf/blob/feature/forgejo-vcs/docs/docs/forgejo_setup.md) on how to set things up. I don't know how that should be organized, but I hope this is a starting point.

I think this looks very comprehensive, it'll do nicely.

Infinoid commented:
> Indeed it is the reporter that calls `SetStatus()`. Perhaps you can add some logging there? As for why it's not being called, I don't know, but it's worth checking the logic of the reporter's `handleRun()` method and its unit tests. Again, adding further logging in the reporter may help work out what's going on.

Good idea.  Okay.  I added some logging, and it skips reporting because it thinks the events came from an API call?

[Here's](https://github.com/leg100/otf/blob/master/internal/run/reporter.go#L82-L84) the check that gets out early.

I added these log messages (among others):

```go
	for event := range sub {
		r.V(0).Info("Reporter got event", "type", event.Type)
[…]
func (r *Reporter) handleRun(ctx context.Context, run *Run) error {
	r.V(0).Info("Reporter.handleRun: called")
	// Skip runs triggered via the UI or API
	if run.Source == SourceUI || run.Source == SourceAPI {
		r.V(0).Info("skipping out: bad source", "source", run.Source)
		return nil
	}
```

Then I opened a PR.  OTF received the webhook, grabbed the tarball, and ran a plan.  I got these logs along the way:

```
2025/05/19 23:34:46 creating forgejo client object
2025/05/19 23:34:46 forgejo.GetRepoTarball(tf/test) called
2025/05/19 23:34:46 ref is c04ccc3e9701912f84f5ae71c108ee06800b3e7a, fqref is c04ccc3e9701912f84f5ae71c108ee06800b3e7a

2025/05/19 23:34:46 INFO Reporter got event component=reporter type=created
2025/05/19 23:34:46 INFO Reporter.handleRun: called component=reporter
2025/05/19 23:34:46 INFO skipping out: bad source component=reporter source=tfe-api
                                                                     ^^^^^^^^^^^^^^

2025/05/19 23:34:46 INFO enqueued plan id=run-c4xvQRDZ7L4vEb5u

2025/05/19 23:34:46 INFO Reporter got event component=reporter type=updated
2025/05/19 23:34:46 INFO Reporter.handleRun: called component=reporter
2025/05/19 23:34:46 INFO skipping out: bad source component=reporter source=tfe-api

2025/05/19 23:34:46 INFO allocated job job.job_id=job-mvMZSdoJRQDXTVgY job.run_id=run-c4xvQRDZ7L4vEb5u job.workspace_id=ws-6AvLSdPG9QnMnKsW job.organization=example-com job.phase=plan job.status=allocated runner_id=runner-11ekoeLxqZJNhbsd
2025/05/19 23:34:46 INFO started plan id=run-c4xvQRDZ7L4vEb5u
2025/05/19 23:34:47 INFO started job job.job_id=job-mvMZSdoJRQDXTVgY job.run_id=run-c4xvQRDZ7L4vEb5u job.workspace_id=ws-6AvLSdPG9QnMnKsW job.organization=example-com job.phase=plan job.status=allocated

2025/05/19 23:34:47 INFO Reporter got event component=reporter type=updated
2025/05/19 23:34:47 INFO Reporter.handleRun: called component=reporter
2025/05/19 23:34:47 INFO skipping out: bad source component=reporter source=tfe-api

2025/05/19 23:34:54 INFO Reporter got event component=reporter type=updated
2025/05/19 23:34:54 INFO Reporter.handleRun: called component=reporter
2025/05/19 23:34:54 INFO skipping out: bad source component=reporter source=tfe-api

2025/05/19 23:34:54 INFO finished job successfully job.job_id=job-mvMZSdoJRQDXTVgY job.run_id=run-c4xvQRDZ7L4vEb5u job.workspace_id=ws-6AvLSdPG9QnMnKsW job.organization=example-com job.phase=plan job.status=allocated
2025/05/19 23:34:54 INFO finished plan id=run-c4xvQRDZ7L4vEb5u resource_changes=+0/~1/−0 output_changes=+0/~0/−0 run_status=planned_and_finished

2025/05/19 23:34:54 INFO Reporter got event component=reporter type=updated
2025/05/19 23:34:54 INFO Reporter.handleRun: called component=reporter
2025/05/19 23:34:54 INFO skipping out: bad source component=reporter source=tfe-api
```

So... what's so bad about `tfe-api`?  Is it supposed to represent user-initiated API calls?  (I didn't call any APIs, I just parsed the webhook event and OTF did its thing.)

Is this a preexisting bug, or something I broke?  Or does something need adding, to recognize events coming from forgejo?

Infinoid commented:
I was able to answer my own question; it's a missing switchcase in [here](https://github.com/leg100/otf/blob/master/internal/run/spawner.go#L216).

I added it, and now the plan result shows up in the PR.  Awesome.

![image](https://github.com/user-attachments/assets/b3b72249-189d-4470-b6e9-47c9fe36565d)

This was concerning, though... there are many places where other similar bugs could be hiding.

leg100 commented:
> I was able to answer my own question; it's a missing switchcase in [here](https://github.com/leg100/otf/blob/master/internal/run/spawner.go#L216).

Yes, that's a code smell and in dire need of a refactor; it's not your fault and reinforces what I said at the beginning that adding a VCS provider requires altering code in lots of different places, and it shouldn't.

leg100 commented:
> I added it, and now the plan result shows up in the PR. Awesome.

Great stuff!

Infinoid commented:
I think this is ready to go.  Please let me know if there's a good way to split it up for ease of review.

leg100 commented:
@Infinoid It looks pretty perfect. I might add an integration test to top it off.

leg100 commented:
@Infinoid I've renumbered the database migration, because I've added a couple on master the last few days.
